pyogrio/geopandas.py

```diff
@@ -326,8 +353,23 @@ def write_dataframe(
     # TODO: may need to fill in pd.NA, etc
     field_data = []
     field_mask = []
+    # dict[str, np.array(datetime.datetime)] special case for dt-tz fields
+    timezone_cols_metadata = {}
     for name in fields:
-        col = df[name].values
+        ser = df[name]
+        col = ser.values
+        if isinstance(ser.dtype, pd.DatetimeTZDtype):
+            # Deal with datetimes with timezones by passing down timezone separately
+            # pass down naive datetime
+            col = ser.dt.tz_localize(None).values
+            # pandas only supports a single offset per column
+            # access via array since we want a numpy array not a series
+            # (only care about the utc offset, not actually the date)
+            # but ser.array.timetz won't have valid utc offset for pytz time zones
+            # (per https://docs.python.org/3/library/datetime.html#datetime.time.utcoffset) # noqa
```

Comment:
this is probably a bit verbose right now.
Basic idea of this is that I can't seem to find a way to produce a series/ numpy array of timezone offsets in a vectorised way through pandas.

Best I could seem to do is that `ser.array.to_pydatetime()` uses the cython function `ints_to_pydatetime`. Note the DateTimeArray is marked as experimental and doesn't actually document any methods on the website which isn't great. The equivalent Series / DateTimeIndex method is another option, but it throws unavoidable userwarnings for changing in behaviour.

This seems wasteful in terms of duplication (basically the same datetime information twice), but seemed better to pass this down and compute the offsets in cython (but I haven't profiled it either).

Reply:
A potentially faster alternative:

```
In [20]: arr = pd.date_range("2012-10-25 09:00", periods=5, freq="D", tz="Europe/Brussels")

In [21]: arr.tz_localize(None) - arr.tz_convert("UTC").tz_localize(None)
Out[21]: 
TimedeltaIndex(['0 days 02:00:00', '0 days 02:00:00', '0 days 02:00:00',
                '0 days 01:00:00', '0 days 01:00:00'],
               dtype='timedelta64[ns]', freq=None)

```
Although it has some more steps, it's vectorized and avoids creating python datetime.datetime objects, and based on a quick test is much faster. And I _think_ it should give the same result?

Reply:
Thanks, this looks much nicer! We can also do the conversion into the gdal representation directly rather than in cython. I've pushed a commit which should implement this. I will do some performance comparisons but haven't got to that yet.